src/main/scala/mesosphere/marathon/tasks/PortsMatcher.scala

```diff
@@ -30,153 +29,171 @@ case class PortsMatch(hostPortsWithRole: Seq[Option[PortWithRole]]) {
+
+        mappings.iterator.map {
+          case Mapping(_, Some(hostPort)) if hostPort == 0 =>
+            if (!availablePortsWithoutStaticHostPorts.hasNext) {
               log.info(s"Offer [${offer.getId.getValue}]. $resourceSelector. " +
-                s"Cannot find range with host port $hostPort for run spec [${runSpec.id}]")
+                s"Insufficient ports in offer for ${name}")
               None
-          }
-        case PortMapping(_, None, _, _, _, _) =>
-          // None has special meaning in this context: it stops the allocation process. this is a problem
-          // if there's an optional host port in the middle of some mappings list. so instead of None we
-          // generate Some(RequestNone) to indicate that we're not requesting a host port, but there may
-          // still be host ports to allocate so don't stop iterating through the list.
-          Some(RequestNone)
+            } else {
+              Option(availablePortsWithoutStaticHostPorts.next())
+            }
+          case Mapping(_, Some(hostPort)) =>
+            offeredPortRanges.find(_.contains(hostPort)) match {
+              case Some(PortRange(role, _, _, reservation)) =>
+                Some(PortWithRole(role, hostPort, reservation))
+              case None =>
+                log.info(s"Offer [${offer.getId.getValue}]. $resourceSelector. " +
+                  s"Cannot find range with host port $hostPort for ${name}")
+                None
+            }
+          case Mapping(_, None) =>
+            // None has special meaning in this context: it stops the allocation process. this is a problem
+            // if there's an optional host port in the middle of some mappings list. so instead of None we
+            // generate Some(RequestNone) to indicate that we're not requesting a host port, but there may
+            // still be host ports to allocate so don't stop iterating through the list.
+            Some(RequestNone)
+        }
       }
     }
-  }
 
-  /**
-    * Takes `expectedSize` ports from the given iterator if possible. Stops when encountering the first `None` port.
-    */
-  private[this] def takeEnoughPortsOrNone[T <: Request](
-    expectedSize: Int)(ports: Iterator[Option[T]]): Option[Seq[Option[PortWithRole]]] = {
-    val allocatedPorts = ports.takeWhile(_.isDefined).take(expectedSize).flatten.toVector
-    if (allocatedPorts.size == expectedSize)
-      Some(allocatedPorts.map {
-        case RequestNone => None
-        case pr: PortWithRole => Some(pr)
-      })
-    else None
-  }
+    /**
+      * Takes `expectedSize` ports from the given iterator if possible. Stops when encountering the first `None` port.
+      */
+    def takeEnoughPortsOrNone[T <: Request](
+      expectedSize: Int)(ports: Iterator[Option[T]]): Option[Seq[Option[PortWithRole]]] = {
+      val allocatedPorts = ports.takeWhile(_.isDefined).take(expectedSize).flatten.toVector
```

Comment:
I feel like this can be reduced in total collection manipulations- also prefer the generalized "toIndexedSeq" over toVector, but collection.breakOut with a type annotation should be sufficient.